TASK-55: Process tfjs model output

cristian-aldea opened the pull request:
closes #190 

This PR aims to implement model output processing so that the model can be easily used for #191 and #33 

When you call the model you get back an array of DetectionResult, which make it easy to then display these results using the HTML canvas.

To test this PR:
* login
* Go to `http://localhost:4200/camera` by clicking on the camera button
* You'll see a **temporary** test page that will allow you to run the model against an image and display some boxes around detected objects.
* Press the `Click me to detect!` button
* You should see something like this:
![image](https://user-images.githubusercontent.com/23544999/107153282-ac059a00-693a-11eb-8c3c-ff91fd9da164.png)


You can change the image that the model is running against by changing the URL in the `<img>` element. Keep in mind that, for now, the model can only detect `bottles`, `box`, `food` and `cans` correctly. Try to have an image with the same ratio as the original image I supplied, otherwise, please adjust the width and height variables so that the canvas fully covers the image

cristian-aldea committed:
feat(ml): Implement model output processsing

I've created a TensorflowJsModel class, which adapts a tfjs graph and processes its output before returning it. A test page was setup so that people can try out the detection

ref #190

cristian-aldea committed:
fix(test): Implement new tests for ml model service

Also, fixed all test and lint issues

ref #190

RGPosadas commented on package-lock.json:
```diff
@@ -10328,22 +10493,14 @@
           }
         },
         "graphql-upload": {
```
Weird, this happened in [Ragith's PR](https://github.com/RGPosadas/Mull/pull/200#discussion_r571444083) as well. In my case, running `npm i` reinstalls graphql-upload to 11.0.0.

cristian-aldea replied:
I don't get what the issue is, we have this part in our `package.json`:
```json
 "resolutions": {
    "graphql-upload": "^11.0.0"
  },
```
isn't 11 what we want

RGPosadas replied:
Yes exactly, 11 is what we want. But your PR brings it back down to 8.1.0

cristian-aldea replied:
oh whoops I misread, I did another npm i and I put it back to 11, I'll be pushing that

RGPosadas reviewed (commented):
Played around with it, and ran into this issue. The banana peel was identified, however the annotation name is out of view.
![image](https://user-images.githubusercontent.com/33991121/107151465-b753c800-6930-11eb-9177-c87c0024ba1c.png)

TheGreatMarkus commented:
> Played around with it, and ran into this issue

Interesting, however, I believe that's more an issue with your image, or maybe the model. If the image was more zoomed out you probably would see the label.

The actual processing of the data seems to be working fine. And keep in mind that the boxes and labels you're seeing are all temporary, and will be replaced by icons in the center of the box in #191

cristian-aldea committed:
fix(packages): Install graphql upload 11

Also moved width and height to variables

cristian-aldea commented on ml/trash-recognition/annotations/label_map.pbtxt:
```diff
@@ -14,6 +14,6 @@ item {
 
 item {
-    id: 3
+    id: 4
```
For anyone who's curious, I screwed up the training of the model by putting a 3 here instead of a 4. This made it so that the model that we're currently using will label cans and tin cans as "food", since they had the same id.

I will be retraining the model soon to fix this, but it shouldn't affect your review, just don't put images of cans for now

RGPosadas reviewed (approved):
Ah true! In that case, what I had mentioned is a non-issue.

> And keep in mind that the boxes and labels you're seeing are all temporary and there so that we can test, and will be replaced by icons in the center of the box in #191 

Great work, LGTM! 🚀 

**Acceptance Criteria**
- [x] Have utility methods and abstractions in place for processing model output
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function

cristian-aldea committed:
fix(ml): Fix wrong order of model output

Initially, we assumed that the order of the tensors within the model output were the same. however, after uploading a new model everything broke. The order of the tensors aren't determined. this was fixed by explicitly asking the model for the tensors we need by name (the names are obfuscated/lost, so a comment clarifies what each tensor is).

cindyslittleplanet reviewed (approved):
LGTM! More training is definitely needed in the future, since the model only detects "perfect" objects. For example, a squeezed plastic bottle is considered a box. But, that's for later :) 

**Acceptance Criteria**
- [x] Have utility methods and abstractions in place for processing model output
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function

ys-lin reviewed (approved):
LGTM :3 
- [x] Have utility methods and abstractions in place for processing model output
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function

TheGreatMarkus commented:
> LGTM! More training is definitely needed in the future, since the model only detects "perfect" objects. For example, a squeezed plastic bottle is considered a box. But, that's for later :)

Indeed, there's definitely room for improvement. I'm happy we've found a process that can allow us to tweak the model and its performance without needing to update or change our application.
